### ISSUE TITLE
Rename Subforms and Update References

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1079]
We also need support for renaming subforms. This will need to:
- rename the .subform file
- rename the corresponding .xml file. Note the recent work in #1067
- update any subform components that reference this subform in .form or .subform files.  UPDATE:  this also needs to be updated in the corresponding .xml files, so that they stay in sync.  It may make more sense to update the XML files only and then regenerate the .form and .subform files based on this.
- UPDATE:  update the references in the open editors?  I think it is probably better to have the users reopen the forms in this case.  The other file updates should trigger a "changed on the filesystem prompt"

The first two parts are done already in #1015, but the third part is more complicated, so I have extracted this into a separate issue.   If we don't have a quick fix for this, we may move on to other tasks.

**Comment by feather812002** [comment]
Hello, Joel
I almost done this , after you remind me : I should fix the subfrom name from the intermiddle xml , rather than from the form or sub-form file, I get big progress on this ,
it almost can work at now , after you rename the subfrom , Moonshine-IDE can auto update this from all form and subfrom files.
I will commit this update on 1015 branch .
So that we can merge it to master on this same branch.

**Comment by feather812002** [comment]
Also I saw some interesting changes when I changed the  inter-middle XML document,
the Moonshine-IDE visual editor will auto update follow the inter-middle xml change .
this is so funny , we don't need close & open form or flex UI again , it will real-time update from the xml .
Very Good!

**Comment by JoelProminic** [comment]
I did some testing, and it initially seemed to work.  However, I started seeing some unexpected behavior as I continued testing.  

I see that the subform name in the DXL is not updated after a rename.  In this case, "SubformA" was renamed to "Subform_a":

```
<subform name="SubformA" version="11.0" maintenanceversion="1.4" xmlns="http://www.lotus.com/dxl">
```

As a result, the updated `<subformref>` XML matched the file name, but not the actual subform name.

In other cases, I saw that `<subformref>` did not update properly for a for a form.  However, this still worked in a build because it matched the original subform name.

I am not clear why `<subformref>` was updated in some cases but not others.  I'll update the issue if I find a consistent test procedure for this.

**Comment by feather812002** [comment]
I to do some test in today , I think I almost understand this bug . 
If user rename the subfrom , for current logic , it only update the name from intermiddle xml , 
So, the dxl form or dxl subfrom all not update, but if you open a form or subfrom with Moonshine-IDE Visual Editor 
it will auto load the latest xml and generate the correct dxl , but if you not open it with Visual Editor, the dxl must store the old subform name  ,so maybe some time you will got a wrong UI .

**Comment by JoelProminic** [comment]
1. Open fresh copy of TestSubform
2. Create a new subform:  TestRename
3. Add a paragraph, label, and field to the subform and save
4. Add the subform to SubformB.subform and TestSubformNesting.form
5. Rename the subform to TestRename2
6. Rename SubformB to Subform_b
7. Rename SubformA to Subform_a
8. Rename TestRename2 to TestRename3
9. Project > Build with Apache Maven
10.  Open the database in the Notes client and confirm it looks right.

In my latest test, this did trigger any errors, and the rename logic looks good overall.  However, during the meeting I was seeing inconsistent errors including:
- Subforms not displaying in form (usually for SubformA)
- Duplicate subformref entries after a rename, some with the old name.

You can use these commands to quickly check the subform references:

    # nsfs/nsf-moonshine
    grep -r "<subformref" .
    # visualeditor-src - I found it easier to keep this separate, but you could add the -i option to see both
    grep -r "<Subformref" .

**Comment by feather812002** [comment]
@JoelProminic 
Hello, Joel , I test this again in today ,follow your test steps.
But it look like it work fine .

I not see the "SubfromA" list in the command search :
![image](https://user-images.githubusercontent.com/8543391/187680958-7c8ed2d5-06c3-41a7-b691-ed62ca1d2b54.png)

I also not see the duplication subform name .

So , maybe we can move to other task , and just keep an eye on this .

**Comment by JoelProminic** [comment]
There were some remaining errors for Subform rename that were discussed resolved in #1015
